docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,164 @@
+# SDG API - Add a file format for defining custom Flows
+
+## Problem Statement
+
+The `instructlab/sdg` library is introducing more extensive data generation pipelines. To enable customization, we should allow users of the library to provide a configuration file which defines a custom pipeline or extends an existing pipeline.
+
+In terms of the API constructs, a Pipeline is created from a sequence of “block configurations” which express how to instantiate and invoke the individual steps (aka blocks) in the pipeline. A Flow construct serves as a template from which a sequence of block configs can be generated.
+
+## Objective
+
+- Library users can specify a custom flow using a well-defined file format.
+- Library users can either use a custom flow standalone, or combine a custom flow with existing flows.
+- The file format and library can evolve substantially without making breaking changes.
+- Incompatible changes can be introduced while retaining support for existing custom flows for a deprecation period.
+
+## Proposal
+
+We will add an API that instantiates a `Pipeline` object from a list of files:
+
+```python
+pipeline = Pipeline.from_flows(ctx, ['mycustomflow.yaml'])
+```
+
+The YAML file format will mirror the API and look like this:
+
+```yaml
+version: 1.0
+blocks:
+  - name: gen_knowledge
+    type: LLMBlock
+    config: # LLMBlock constructor kwargs
+      output_cols: ["question", "response"]
+    gen_kwargs: # kwargs for block.generate()
+      max_tokens": 2048,
+    drop_duplicates: ["question"]
+  - name: filter_faithfulness
+    type: FilterByValueBlock
+    config:
+      filter_column: judgment
+      filter_value: YES
+      operation: eq
+      drop_columns: ["judgment", "explanation"]
+```
+
+## Versioning
+
+A mandatory `version` field in the YAML file expresses major and minor versions (e.g., 1.0, 1.1, 2.0).
+
+Compatibility rules
+
+1. If the major version of the YAML file is higher than the parser can handle, the parser should reject the file.
+2. If the minor version of the YAML file is higher than the highest version the parser is aware of, the parser should read the file but ignore any unrecognized content.
+3. If the file’s version is lower than the parser version, the parser should provide default values for any configuration introduced in later versions.
+
+Example parsing logic:
+
+```python
+def parse_custom_flow(content):
+    version = content['version']
+    major, minor = map(int, version.split('.'))
+
+    if major > PARSER_MAJOR:
+        raise IncompatibleVersionError("The custom flow file format is from a future major version.")
+    elif major <= PARSER_MAJOR and minor > PARSER_MINOR:
+        logger.warning("The custom flow file may have new features that will be ignored.")
+```
+
+### Flow Params
+
+The following parameters are supplied to a Flow constructor and used by the render() method:
```

Comment:
i don't think client should be part of the config, nor num_instructions_to_generate and batched.
why? because the same flow should be triggered at runtime with a flexible combination of those

Reply:
in other words, those are running details but don't affect the "semantics" of the flow.

Reply:
Absolutely :+1: 

I've started referring to these as "runtime parameters", and this is why I've proposed the `PipelineContext` object

When I say "The following parameters are supplied to a Flow constructor", I'm referring to the current code, not what I'm proposing

But I'm not being very clear here. I will update :+1: